### DISCUSSION TITLE
Questions about coverage

### DISCUSSION THREAD

**Discussion by tmeissner** [nickg/nvc#888] [Q&A] [answered]
I have two questions about coverage.

### Can I set the name and folder of the coverage database file?
At the moment nvc writes it in the _work directory with a name like `work/_WORK.QUEUEHANDLINGT.elab.covdb`.

### How to avoid the ` Could not locate source file: (null) that NVC used to collect coverage for` warnings?

I get warnings like this when I process the coverage data:

```
nvc -c --report=coverage work/_WORK.QUEUEHANDLINGT.elab.covdb
** Note: Code coverage report folder: coverage.
** Note: Code coverage report contains: covered, uncovered, excluded coverage details.
** Warning: Could not locate source file: (null) that NVC used to collect coverage for: WORK.QUEUEHANDLINGT.H2CAXIREADREQP._L0. Dropping coverage for this hierarchy/item from coverage report.
** Warning: Could not locate source file: (null) that NVC used to collect coverage for: WORK.QUEUEHANDLINGT.C2HAXIREADREQP._L0. Dropping coverage for this hierarchy/item from coverage report.
** Warning: Could not locate source file: (null) that NVC used to collect coverage for: WORK.QUEUEHANDLINGT.C2HAXIWRITEREQP._L0. Dropping coverage for this hierarchy/item from coverage report.
** Warning: Could not locate source file: (null) that NVC used to collect coverage for: WORK.QUEUEHANDLINGT.C2HSTIMULIP._L0. Dropping coverage for this hierarchy/item from coverage report.
** Warning: Could not locate source file: (null) that NVC used to collect coverage for: WORK.QUEUEHANDLINGT.CMPSTIMULIP._L0. Dropping coverage for this hierarchy/item from coverage report.
** Warning: Could not locate source file: (null) that NVC used to collect coverage for: WORK.QUEUEHANDLINGT.QUEUECONTEXTP._L0. Dropping coverage for this hierarchy/item from coverage report.
** Note: code coverage results for: WORK.QUEUEHANDLINGT
```

**Comment by nickg** [comment]
> Can I set the name and folder of the coverage database file?

Not at the moment but I think it's a good idea to add. And it should probably default to something in the working directory rather than the library directory.

> How to avoid the  Could not locate source file: (null) that NVC used to collect coverage for warnings?

This is most likely a bug where source location is not tracked correctly for some elements. `WORK.QUEUEHANDLINGT.H2CAXIREADREQP._L0` is probably the first loop inside process `H2CAXIREADREQP`. Could you see what kind of construct that is (e.g. while loop, for loop, etc.)?

**Reply by nickg** [reply]
I managed to reproduce this with a `loop ... end loop;` statement. It should be fixed in 04aa33d. Could you try again with the latest master branch?

**Reply by tmeissner** [reply]
Ui, that was fast 😅 I will try tomorrow at work. And, yes, I think there are loops in the processed which call some AXI BFM functions until Test is finished.

**Reply by tmeissner** [reply]
I can confirm that it works now without the warnings. Thanks 🙂

**Comment by tmeissner** [comment, marked as answer]
Another question regarding coverage:

I don't understand how to correctly use the coverage setting files. I only want coverage from my DUT and the hierarchy under it, but not the test bench itself. My test bench name is `QUEUEHANDLINGT`, so I tried something like this:

```
- block QUEUEHANDLINGT
+ hierarchy WORK.QUEUEHANDLINGT.*
```

But then coverage is empty. When I don't use these settings I get coverage, but also for the test bench which I don't want. So, how can I exclude coverage of the (top level) test bench, but including coverage of the DUT and all units "under" it?

**Reply by nickg** [reply]
I think it should work with just a single `hierarchy` command like:

```
+hierarchy WORK.QUEUEHANDLINGT.UUT.*
```

Because the default is to exclude if the spec file is present. @Blebowski

**Reply by Blebowski** [reply]
Yup, thats the way it was intended. If you specify the spec file, you need to enable the block / hierarchy that you want to collect. Default is "nothing is collected". If this was done in reverse, in your situation (DUT in TB), you would need to disable TB (that would by transition disable DUT), so you would need to enable collection for DUT again. Since disable has priority over enable, you would again end up in empty coverage database,

The goal was to achieve following scenario:
  TB_TOP
  ----DUT
  --------DATAPATH
  ------------FETCH_STAGE
  ------------MEMORY_CONTROLLER
  ----------------MEMORY_MACRO_BEHAVIORAL_MODEL
  ------------DECODE_STAGE
  ------------EXECUTE_STAGE
  --------CONTROLLER
  ------------SOME_FSM_CONTROLLING_THE_CPU

You have TB_TOP, under it is some DUT, and within the DUT you want to collect everything with some exceptions (e.g. the MEMORY_MACRO_BEHAVIORAL_MODEL is non-synthesizable code, so you don't collect coverage there).
With current implementation, you simply do:

```
+ hierarchy TB_TOP.DUT
-  hierarchy TB_TOP.DUT.DATAPATH.MEMORY_CONTROLLER.MEMORY_MACRO_BEHAVIORAL_MODEL
```

The `hierarchy` works based on instance name and its complete path. The `block` works for any entity with the
given name. So by saying e.g.:
`-block CLOCK_GATER`
You would disable collection on all instances of entity `CLOCK_GATER`. Assuming clock_gater is only RTL model,
and for synthesis you replace this by some technology specific implementation, you can get rid of uncovered stuff
without manual waiving quite quickly.
This is usefull when you have some library of common cells (e.g. FIFOs, clock gaters, synthesizable RAMs, various
decoders and converters, gate primitives), and this library has generic code with lot of parametrization, so you
don't want to waive each unreachable coverage data manually, instead you avoid collecting the coverage on each
entity from such library. Assuming the library is already verified...

**Reply by tmeissner** [reply]
I tried and this seems to work:
```
+hierarchy WORK.QUEUEHANDLINGT.UUT.*
```
I have to add the `.*` after the UUT, without it, no coverage is collected.

Thanks for the explanation 👍

**Reply by Blebowski** [reply]
Of sorry, I forgot about the wildcard, yes, it the hierarchy name is matched by wildcard the same as waveform dumping.
